feat(ir): add list comprehension to IR

tohrnii opened the pull request:
Partly addressing #87. This PR adds list comprehension and list folding support to the IR.

bobbinth commented on ir/src/constraints/list_comprehension.rs:
```diff
@@ -0,0 +1,328 @@
+/// Returns an error if there is an error while parsing the sub-expression.
+fn parse_lc_expr(
+    expression: &Expression,
+    lc: &ListComprehension,
```
It seems to me that we are passing `ListComprehension` around to get access to its context and then use loops to figure out which member corresponds to which iterable. If so, a better solution could be to build a map from identifier to iterable upfront (maybe put this map into something a struct - something like `IterableContext`), and then pass this map around instead `ListComprehension`.

This way, instead of doing something like:
```Rust
for (member, iterable) in lc.context() {
    if ident == member {
        match iterable {
```
You'd be able to do something like:
```Rust
let iterable = context.get_iterable(ident)?;
match iterable {
```

You may be able to take it even a step further and put `IdentifierType` for each member into the context. This way, you won't need to pass `symbol_table` around either.

tohrnii replied:
Thanks, done. I moved it to just a `BTreeMap<Identifier, Iterable>`. I didn't remove symbol table as it is required to build context for list folding inside list comprehension.

bobbinth commented on air-script/tests/list_comprehension/list_comprehension.air:
```diff
@@ -0,0 +1,29 @@
+    enf c[2].first = 0
+
+integrity_constraints:
+    let x = [fmp for fmp in fmp]
+    enf clk = x[1]
+    let enumerate = [2^i * c for (i, c) in (0..4, c)]
+    enf clk = enumerate[3]
+    let diff0 = [x - y for (x, y) in (c, d)]
+    enf clk = diff0[0]
+    let diff1 = [x - y for (x, y) in (c[0..2], d[1..3])]
+    enf clk = diff1[1]
+    let x0 = [w + x - y - z for (w, x, y, z) in (0..3, b, c[0..3], d[0..3])]
+    enf a = x0[0] + x0[1] + x0[2]
+    let x1 = sum([c * d for (c, d) in (c, d)])
+    let y1 = prod([c + d for (c, d) in (c, d)])    
+    enf clk = x1
+    let x2 = sum([c * d for (c, d) in (c, d)])
+    let y2 = [m + x2 for m in fmp]
+    enf clk = y2[0]
```
minor nit: I'd probably add blank lines after each `enf` - would make it a bit easier to read.

Also, could we add a test for using "next" value from a column?

bobbinth commented on ir/src/constraints/list_comprehension.rs:
```diff
@@ -0,0 +1,328 @@
+                                ))?,
+                            }
+                        }
+                        IdentifierType::PublicInput(size) => {
+                            validate_access(i, *size)?;
+                            return Ok(Expression::VectorAccess(VectorAccess::new(
+                                ident.clone(),
+                                i,
+                            )));
+                        }
```
Public inputs should not be accessible in integrity constraints. Is this checked somewhere later?

tohrnii replied:
Thanks for catching this. It's not checked currently. I've added a check later on to throw an error if public inputs are accessed in integrity constraints.

bobbinth commented on ir/src/constraints/list_comprehension.rs:
```diff
@@ -0,0 +1,328 @@
+                    match ident_type {
+                        IdentifierType::TraceColumns(trace_columns) => {
+                            validate_access(i, trace_columns.size())?;
+                            return Ok(Expression::NamedTraceAccess(NamedTraceAccess::new(
+                                ident.clone(),
+                                i,
+                                CURRENT_ROW,
```
Why do we hard-code `CURRENT_ROW` here? Shouldn't the offset come from `trace_columns`?

tohrnii replied:
The AST has current row trace columns stored as general element since there's no way to disambiguate between whether it's a trace column or any other symbol table like an intermediate variable. So, if it's stored as a trace column in symbol table, we know that it should have 0 row offset here.

bobbinth commented on ir/src/constraints/list_comprehension.rs:
```diff
@@ -0,0 +1,328 @@
+                    match ident_type {
+                        IdentifierType::TraceColumns(trace_columns) => {
+                            validate_access(i, trace_columns.size())?;
+                            return Ok(Expression::NamedTraceAccess(NamedTraceAccess::new(
+                                ident.clone(),
+                                range.start() + i,
+                                CURRENT_ROW,
+                            )));
```
Why do we hard-code `CURRENT_ROW` here? Shouldn't the offset come from `trace_columns`?

tohrnii replied:
The AST has current row trace columns stored as general element since there's no way to disambiguate between whether it's a trace column or any other symbol table like an intermediate variable. So, if it's stored as a trace column in symbol table, we know that it should have 0 row offset here.

bobbinth replied:
I might be missing something here, but how is something like this stored in the symbol table:
```
trace_columns:
    main: [a, b]

integrity_constraints:
    let x = a'
    enf x = b + 2
```

tohrnii replied:
The trace columns are stored as:
```
TraceColumns(TraceColumns)
```
where
```
pub struct TraceColumns {
    trace_segment: TraceSegment,
    offset: usize,
    size: usize,
}
```
where offset is the column offset of the trace column group instead of row offset.

When we insert the intermediate variable `let x = a'`, `a'` would be stored in the symbol table as:
```
TraceElement(IndexedTraceAccess { trace_segment: 0, col_idx: 0, row_offset: 1 })
```

I think the confusion might be that `a` and `a'` are stored differently in the AST. When we encounter `a` at the parser level we don't know if it's a trace column or any other identifier so we store it as `Expression::Elem`, however when we encounter `a'` at the parser level, we know that it should be referring it to a trace column so we store it as `Expression::NamedTraceAccess(NamedTraceAccess)` where
```
pub struct NamedTraceAccess {
    name: Identifier,
    idx: usize,
    row_offset: usize,
}
```

Maybe there's a better way to do this since this is certainly a little confusing. Please let me know if you have other suggestion.

So, when we parse the AST expression while unfolding list comprehension, if the expression is of type `Elem`, we know that it should be referring to current row.

bobbinth commented on ir/src/constraints/list_comprehension.rs:
```diff
@@ -0,0 +1,328 @@
+                                ))?,
+                            }
+                        }
+                        IdentifierType::PublicInput(size) => {
+                            validate_access(i, *size)?;
+                            return Ok(Expression::VectorAccess(VectorAccess::new(
+                                ident.clone(),
+                                range.start() + i,
+                            )));
+                        }
```
Public inputs should not be accessible in integrity constraints. Is this checked somewhere later?

bobbinth commented on ir/src/constraints/list_comprehension.rs:
```diff
@@ -0,0 +1,328 @@
+                    match ident_type {
+                        IdentifierType::TraceColumns(size) => {
+                            validate_access(i, size.size())?;
+                            return Ok(Expression::NamedTraceAccess(NamedTraceAccess::new(
+                                ident.clone(),
+                                i,
+                                NEXT_ROW,
+                            )));
```
Why do we hard-code `NEXT_ROW` here? Shouldn't the offset come from `trace_columns`?

tohrnii replied:
You are right, it should come from `named_trace_access` here. Thanks.

bobbinth commented on ir/src/constraints/list_comprehension.rs:
```diff
@@ -0,0 +1,328 @@
+                    match ident_type {
+                        IdentifierType::TraceColumns(trace_columns) => {
+                            validate_access(i, trace_columns.size())?;
+                            return Ok(Expression::NamedTraceAccess(NamedTraceAccess::new(
+                                ident.clone(),
+                                range.start() + i,
+                                NEXT_ROW,
+                            )));
```
Why do we hard-code `NEXT_ROW` here? Shouldn't the offset come from `trace_columns`?

tohrnii replied:
You are right, it should come from `named_trace_access` here. Thanks.

bobbinth commented on ir/src/constraints/graph.rs:
```diff
@@ -166,7 +173,38 @@ impl AlgebraicGraph {
-            Expression::ListFolding(_) => todo!(),
+            Expression::ListFolding(lf_type) => match lf_type {
+                ListFoldingType::Sum(lc) => {
+                    let list = parse_lc(lc, symbol_table)?;
+                    let mut sum =
+                        self.insert_expr(symbol_table, &list[0], variable_roots, default_domain)?;
+                    for elem in list.iter().skip(1) {
+                        let expr =
+                            self.insert_expr(symbol_table, elem, variable_roots, default_domain)?;
+                        sum = self.insert_bin_op(
+                            &sum,
+                            &expr,
+                            Operation::Add(sum.root_idx(), expr.root_idx()),
+                        )?;
+                    }
+                    Ok(sum)
```
It may be possible to create a function which does this for both `Sum` and `Prod` - the only thing which is different between the two is which binary operation we use in the end.

tohrnii replied:
Done

bobbinth reviewed (changes requested):
Thank you! Looks good! This is not a full review yet - but I left some comments inline.

tohrnii commented on ir/src/constraints/list_comprehension.rs:
```diff
@@ -0,0 +1,328 @@
+                        "List comprehensions only allowed for vector variables.".to_string(),
+                    )),
+                },
+                IdentifierType::PublicInput(size) => Ok(*size),
```
Thanks for catching this. It's not checked currently. I've added a check later on to throw an error if public inputs are accessed in integrity constraints.

bobbinth reviewed (changes requested):
Thank you! Still not a full review - but I left some comments inline.

I think the biggest issue is that the constraint test file (`list_comprehension.air`) seems to be compiled incorrectly into Rust code. I'm not sure if this is due to this PR or is a prior issue - but I think it would be difficult to check the behavior until this is corrected.

bobbinth commented on air-script/tests/list_comprehension/list_comprehension.rs:
```diff
@@ -0,0 +1,94 @@
+    {
+        let current = aux_frame.current();
+        let next = aux_frame.next();
+        result[0] = current[0] - (((E::from(2_u64)).exp(E::PositiveInteger::from(3_u64))) * (current[7]));
```
Actually, I would even say that the expression should have been just `E::from(8_u64)` but maybe this is an optimization step.

tohrnii replied:
Thanks. Yeah, I think the second one should probably be an optimization step. I'll add an issue to track these.

tohrnii commented on ir/src/constraints/list_comprehension.rs:
```diff
@@ -0,0 +1,350 @@
+    symbol_table: &SymbolTable,
+    i: usize,
+) -> Result<Expression, SemanticError> {
+    let iterable = iterable_context.get(ident);
+    if let Some(iterable_type) = iterable {
+        match iterable_type {
```
@bobbinth I think we can have symbols in a list comprehension expression that are not in the iterables of the list comprehension. For example, the following should be valid:
```
trace_columns:
    main: [c[3]]
let x = 1
let a = [b + x for b in c]
```
Here, when we are unrolling the list comprehension, we just leave x alone and return it if we don't find a corresponding iterable. So the above list comprehension unrolls into:
```
[c[0] + x, c[1] + x, c[2] + x]
```

bobbinth replied:
Ah - makes sense! I'd probably add some comments and maybe restructure the function to make it more clear.

For example, for `parse_elem()` do we really need to pass in `expression`? If I understood it correctly, passing both `ident` and `expression` is redundant. And the only reason we pass it in is to clone it at the end. But we could have just as easily done something like `Expression::Elem(ident)` right?

And the logic of the whole function could be:
1. If the `ident` is not in the `iterable_context`, then just return `Expression::Elem(ident)` (and I guess any invalid references will be caught later).
2. Otherwise, process the `ident` as we do now.

tohrnii commented on ir/src/constraints/mod.rs:
```diff
@@ -232,7 +236,15 @@ impl Constraints {
             IntegrityStmt::Variable(variable) => {
-                symbol_table.insert_integrity_variable(variable)?
+                if let VariableType::ListComprehension(list_comprehension) = variable.value() {
+                    let vector = unfold_lc(list_comprehension, symbol_table)?;
```
`expressions` seems a little vague to me since we are returning a vector of expressions to be used for a vector type variable in this context. Do you think `variable_value` is more accurate? Not a strong opinion though, I'm happy to go with `expressions` if that seems better to you.

grjte replied:
I probably prefer `expressions` as well, but you could use `expr_vector` to capture both ideas

tohrnii commented on air-script/tests/list_comprehension/list_comprehension.air:
```diff
@@ -0,0 +1,35 @@
+    let x = [fmp for fmp in fmp]
+    enf clk = x[1]
+
+    let enumerate = [2^i * c for (i, c) in (0..4, c)]
+    enf clk = enumerate[3]
+    
+    let diff0 = [x' - y' for (x, y) in (c, d)]
+    enf clk = diff0[0]
+    
+    let diff1 = [x - y for (x, y) in (c[0..2], d[1..3])]
+    enf clk = diff1[1]
+    
+    let x0 = [w + x - y - z for (w, x, y, z) in (0..3, b, c[0..3], d[0..3])]
+    enf a = x0[0] + x0[1] + x0[2]
+    
+    let x1 = sum([c * d for (c, d) in (c, d)])
+    let y1 = prod([c + d for (c, d) in (c, d)])    
+    enf clk = x1
+    
+    let x2 = sum([c * d for (c, d) in (c, d)])
+    let y2 = [m + x2 for m in fmp]
+    enf clk = y2[0]
```
Makes sense. I was a little sloppy in writing these examples (just wanted to write down different kinds of expressions possible. I'll improve on them to only include valid expressions and will add some useful ones.

tohrnii commented on air-script/tests/list_comprehension/list_comprehension.rs:
```diff
@@ -0,0 +1,94 @@
+        result[0] = current[0] - (current[2]);
+    }
+
+    fn evaluate_aux_transition<F, E>(&self, main_frame: &EvaluationFrame<F>, aux_frame: &EvaluationFrame<E>, _periodic_values: &[F], aux_rand_elements: &AuxTraceRandElements<E>, result: &mut [E])
+    where F: FieldElement<BaseField = Felt>,
+          E: FieldElement<BaseField = Felt> + ExtensionOf<F>,
+    {
+        let current = aux_frame.current();
+        let next = aux_frame.next();
+        result[0] = current[0] - (((E::from(2_u64)).exp(E::PositiveInteger::from(3_u64))) * (current[7]));
+        result[1] = current[0] - (next[4] - (next[8]));
+        result[2] = current[0] - (current[5] - (current[10]));
+        result[3] = current[0] - (E::from(0_u64) + current[1] - (current[4]) - (current[8]) + E::from(1_u64) + current[2] - (current[5]) - (current[9]) + E::from(2_u64) + current[3] - (current[6]) - (current[10]));
+        result[4] = current[0] - ((current[4]) * (current[8]) + (current[5]) * (current[9]) + (current[6]) * (current[10]) + (current[7]) * (current[11]));
+        result[5] = current[0] - (current[1] + (current[4]) * (current[8]) + (current[5]) * (current[9]) + (current[6]) * (current[10]) + (current[7]) * (current[11]));
+    }
```
Thanks for catching this. This seems like it was an existing issue. I'll open a PR today with a fix.

tohrnii replied:
I've opened #148.

bobbinth commented on ir/src/constraints/list_comprehension.rs:
```diff
@@ -0,0 +1,342 @@
+use std::collections::BTreeMap;
+
+use super::{graph::CURRENT_ROW, IdentifierType, SemanticError, SymbolTable};
```
minor nit: I'd remove the blank line.

bobbinth commented on ir/src/constraints/graph.rs:
```diff
@@ -398,6 +423,14 @@ impl AlgebraicGraph {
             IdentifierType::PublicInput(_) => {
+                if !matches!(
+                    domain,
+                    ConstraintDomain::FirstRow | ConstraintDomain::LastRow
```
Should we add `is_boundary()` to `ConstraintDomain`?

tohrnii replied:
Done

bobbinth commented on ir/src/constraints/graph.rs:
```diff
@@ -165,7 +170,27 @@ impl AlgebraicGraph {
                 ))
             }
-            Expression::ListFolding(_) => todo!(),
+            Expression::ListFolding(lf_type) => match lf_type {
+                ListFoldingType::Sum(lc) | ListFoldingType::Prod(lc) => {
+                    let list = unfold_lc(lc, symbol_table)?;
+                    let mut acc =
+                        self.insert_expr(symbol_table, &list[0], variable_roots, default_domain)?;
+                    for elem in list.iter().skip(1) {
+                        let expr =
+                            self.insert_expr(symbol_table, elem, variable_roots, default_domain)?;
+                        let op = match lf_type {
+                            ListFoldingType::Sum(_) => {
+                                Operation::Add(acc.root_idx(), expr.root_idx())
+                            }
+                            ListFoldingType::Prod(_) => {
+                                Operation::Mul(acc.root_idx(), expr.root_idx())
+                            }
+                        };
+                        acc = self.insert_bin_op(&acc, &expr, op)?;
+                    }
+                    Ok(acc)
+                }
```
Let's move this into a separate function - something like `insert_list_folding()` (similar to how we have separate functions `insert_vector_access()` etc.).

bobbinth reviewed (changes requested):
Looks good! Thank you! This is still not a full review - but I left a few comments inline.

tohrnii commented on ir/src/constraints/list_comprehension.rs:
```diff
@@ -0,0 +1,342 @@
+    lc: &ListComprehension,
+    symbol_table: &SymbolTable,
+) -> Result<Vec<Expression>, SemanticError> {
+    let num_iterations = get_num_iterations(lc, symbol_table)?;
```
Thanks. I've gone with enforcing number of iterations greater than 1 but maybe even 1 should be syntactically correct even if it's not very meaningful so I'm not sure what the right approach is here.

grjte replied:
In cases like this, I generally prefer the syntactically correct option, because it's more flexible for the user. I could imagine a case where we have a list comprehension as part of a evaluator which is reused over a variety of sets of imported trace column groups, most of which consist of multiple columns, but one of which consists of 1 column. In this case, things would break for the programmer and mean that they need to define a separate function for that one case, whereas it's something that we could probably handle cleanly (especially once there are more optimizations in the IR), allowing them to have a clearer code pattern

grjte commented on ir/src/constraints/mod.rs:
```diff
@@ -6,6 +6,7 @@ use super::{
 use std::collections::{BTreeMap, BTreeSet};
 
 mod constraint;
+use air_script_core::Variable;
```
nit: this can be imported in `super` above

grjte commented on ir/src/constraints/list_comprehension.rs:
```diff
@@ -0,0 +1,376 @@
+    let num_iterations = get_num_iterations(lc, symbol_table)?;
+    if num_iterations <= 1 {
+        return Err(SemanticError::InvalidListComprehension(
+            "List comprehension must have more than one iteration.".to_string(),
```
nit: I would change this to "List comprehensions must have"

grjte commented on ir/src/constraints/list_comprehension.rs:
```diff
@@ -0,0 +1,376 @@
+/// Parses a list comprehension expression and creates an expression based on the index of the
+/// expression in the list comprehension.
+///
+/// Returns an error if there is an error while parsing the sub-expression.
```
nit: This should have an `# Errors` header

grjte commented on ir/src/constraints/list_comprehension.rs:
```diff
@@ -0,0 +1,376 @@
+
+/// Unfolds a list comprehension into a vector of expressions.
+///
+/// Returns an error if there is an error while parsing any of the expressions in the expanded
+/// vector from the list comprehension.
```
nit: this should have an `# Errors` header

grjte commented on ir/src/constraints/list_comprehension.rs:
```diff
@@ -0,0 +1,376 @@
+    }
+}
+
+/// Validates and returns the length of a list comprehension. Checks that the length of all iterables
+/// in the list comprehension is the same and returns an error if it's not.
```
nit: the error should be separated into an `# Errors` section. There are several other places with the same problem, so I won't comment them anymore, but please check the rest of the doc comments

tohrnii replied:
Makes sense. I generally take the approach of having a separate section if there are more than one error cases but I don't think I've been consistent with that either. I'll add an errors section everywhere.

grjte commented on ir/src/constraints/list_comprehension.rs:
```diff
@@ -0,0 +1,376 @@
+    }
+}
+
+/// Checks if the access index is valid. Returns an error if the access index is greater than
+/// the size of the vector.
+fn validate_access(i: usize, size: usize) -> Result<(), SemanticError> {
+    if i < size {
+        Ok(())
+    } else {
+        Err(SemanticError::InvalidListComprehension(
+            "Invalid access index".to_string(),
+        ))
+    }
+}
```
I think this should be an `IndexOutOfRange` instead, and then the error message could state that it happens in a list comprehension and provide useful details. Wdyt?

tohrnii replied:
Makes sense. Fixed.

grjte commented on ir/src/constraints/list_comprehension.rs:
```diff
@@ -0,0 +1,376 @@
+    }
+}
+
+/// Builds an [IterableContext] from a given list comprehension.
+fn build_iterable_context(lc: &ListComprehension) -> Result<IterableContext, SemanticError> {
```
There's no documentation of `IterableContext` or of `Iterable`, and I think this is starting to get a bit hard to understand. Can you add brief docs for those so that this doc comment will be more helpful?

grjte commented on ir/src/constraints/list_comprehension.rs:
```diff
@@ -0,0 +1,376 @@
+            .insert(member.clone(), iterable.clone())
+            .is_some()
+        {
+            return Err(SemanticError::InvalidListComprehension(
+                "Duplicate member in list comprehension".to_string(),
+            ));
```
I think maybe this error should be more descriptive so it will more be helpful to anyone unfamiliar with the IR internals

tohrnii replied:
I've changed it to also print the member name in the error message.

grjte commented on ir/src/constraints/graph.rs:
```diff
@@ -1,14 +1,16 @@
+    MatrixAccess, SemanticError, SymbolTable, TraceSegment, VariableRoots, VariableType,
+    VectorAccess,
 };
+use air_script_core::ListFoldingType;
```
nit: I'd prefer to import this at the top level of the crate and then import from `super` here.

grjte commented on ir/src/constraints/graph.rs:
```diff
@@ -398,6 +405,11 @@ impl AlgebraicGraph {
                 Ok(ExprDetails::new(node_index, trace_segment, domain))
             }
             IdentifierType::PublicInput(_) => {
+                if !domain.is_boundary() {
```
now that `is_boundary` is defined, it can be updated in `ConstrainedBoundary` as well. I think it makes sense to do this at the same time as defining the method, unless you object

grjte commented on ir/src/constraints/graph.rs:
```diff
@@ -485,6 +497,42 @@ impl AlgebraicGraph {
+    /// Inserts a list folding expression into the graph and returns the resulting expression
+    /// details.
+    ///
+    /// # Errors
+    /// Panics if the list is empty.
+    /// Returns an error if the list cannot be unfolded properly.
```
should we give these bullet points? I think we usually do when there are multiple error cases, right?

grjte commented on ir/src/constraints/list_comprehension.rs:
```diff
@@ -0,0 +1,376 @@
+    i: usize,
+) -> Result<Expression, SemanticError> {
+    match ident_type {
+        IdentifierType::TraceColumns(trace_columns) => {
+            validate_access(i, trace_columns.size())?;
+            Ok(Expression::NamedTraceAccess(NamedTraceAccess::new(
+                ident.clone(),
+                i,
+                CURRENT_ROW,
+            )))
+        }
```
I know it would mean passing the symbol table, but it might be preferable to create an `IndexedTraceAccess` at this point. Since it's the IR we have all the information, and if we create a `NamedTraceAccess` we just have to convert it later.

tohrnii replied:
Makes sense. Done.

grjte reviewed (changes requested):
This is a partial review. I need to spend more time looking at the `list_comprehension` module. 

As a general comment, I think it would be easier to read this file if the doc comments were more detailed and/or provided examples - in most cases, the doc comments explain what's happening in terms of internal terminology that we've chosen or refer to undocumented structs, so it's a bit hard to understand one thing without already understanding everything. Ideally, everything that's happening could be understood by just reading the function doc comments, and I think that's not really the case right now (or not easily)

tohrnii commented:
@grjte I've addressed your comments except adding more detailed docs. I'm working on them now.

One question is if we should rename `unfold_lc` to something like `expand_lc` (or something else) because unfold might be confusing since we also have list folding?

grjte reviewed (changes requested):
Thanks @tohrnii I think it's looking really good. The comments you've added are very helpful. It feels quite clear to read through it at this point. I left several things inline, but they're all minor.

One final note is that I feel like we might be able to improve the pattern for the `list_comprehension` module using a struct + methods instead of all of the independent functions. I think it would be hard to do this now, but that it might be worth a look after #150 (or maybe as part of it)

tohrnii commented on ir/src/constraints/list_comprehension.rs:
```diff
@@ -0,0 +1,397 @@
+                IdentifierType::Variable(_, var_type) => match var_type.value() {
+                    VariableType::Vector(vector) => Ok(vector.len()),
+                    _ => Err(SemanticError::InvalidListComprehension(
+                        "List comprehensions only allowed for vector variables.".to_string(),
```
Thanks. Changed it to 
```
format!("Variable {} should be a vector for a valid list comprehension.", ident.name()),
```
instead.

tohrnii commented on ir/src/tests/list_comprehension.rs:
```diff
@@ -0,0 +1,306 @@
+#[test]
+fn lc_with_const_exp() {
+    let source = "
+    trace_columns:
+        main: [clk, fmp[2], ctx]
+        aux: [a, b, c[4], d[4]]
+    public_inputs:
+        stack_inputs: [16]
+    boundary_constraints:
+        enf c[2].first = 0
+    
+    integrity_constraints:
+        let y = [col^7 for col in c]
+        enf clk = y[1]";
+
+    let parsed = parse(source).expect("Parsing failed");
+    let result = AirIR::new(&parsed);
+    assert!(result.is_ok());
+}
+
+#[test]
+fn lc_with_named_trace_access_const_exp() {
+    let source = "
+    trace_columns:
+        main: [clk, fmp[2], ctx]
+        aux: [a, b, c[4], d[4]]
+    public_inputs:
+        stack_inputs: [16]
+    
+    boundary_constraints:
+        enf c[2].first = 0
+    
+    integrity_constraints:
+        let y = [col'^7 - col for col in c]
+        enf clk = y[1]";
+
+    let parsed = parse(source).expect("Parsing failed");
+    let result = AirIR::new(&parsed);
+    assert!(result.is_ok());
+}
```
Thanks. I think my thinking at the time was that they are treated differently at the parser level so named accordingly. I have now merged both the test cases.

tohrnii commented:
Thank you @bobbinth and @grjte for the reviews. Made suggested changes.

tohrnii commented on ir/src/tests/list_comprehension.rs:
```diff
@@ -0,0 +1,306 @@
+    let parsed = parse(source).expect("Parsing failed");
+    let result = AirIR::new(&parsed);
+    assert!(result.is_err());
+}
```
moved it to `integrity_constraints` below an exp related test.

grjte commented on air-script/tests/list_comprehension/list_comprehension.rs:
```diff
@@ -0,0 +1,96 @@
+use winter_air::{Air, AirContext, Assertion, AuxTraceRandElements, EvaluationFrame, ProofOptions as WinterProofOptions, TransitionConstraintDegree, TraceInfo};
+use winter_math::fields::f64::BaseElement as Felt;
+use winter_math::{ExtensionOf, FieldElement};
+use winter_utils::collections::Vec;
+use winter_utils::{ByteWriter, Serializable};
```
nit: these could be condensed into 2 lines

tohrnii replied:
I'm not sure how to do that. The import function takes a module path and type. So, currently we do:
```
scope.import("winter_utils::collections", "Vec")
```
If I move `collections` to the type argument:
```
scope.import("winter_utils",  "collections::Vec")
```
this generates output:
```
use winter_utils::{collections, ByteWriter, Serializable};
```
This is what I had done earlier to combine the imports before Bobbin pointed out the problem.

Looking at the import constructor it just joins the two strings:
```
pub fn new(path: &str, ty: &str) -> Self {
        Import {
            line: format!("{}::{}", path, ty),
            vis: None,
        }
    }
```
So, will have to look into it more. If you have any ideas on how to achieve this, please let me know otherwise I'll dig into this more later.

grjte replied:
Ah yeah, makes sense. It's not worth any time before the next version, so let's just leave it.

grjte commented on ir/src/constraints/list_comprehension.rs:
```diff
@@ -0,0 +1,399 @@
+                            named_trace_access.row_offset(),
+                        )))
+                    }
+                    _ => Err(SemanticError::InvalidListComprehension(format!(
+                        "Iterable {ident} should be trace columns"
+                    )))?,
+                }
+            }
+            Iterable::Range(_) => Err(SemanticError::InvalidListComprehension(format!(
+                "Iterable cannot be of type Range for named trace access {}",
+                named_trace_access.name()
+            ))),
+            Iterable::Slice(ident, range) => {
+                let ident_type = symbol_table.get_type(ident.name())?;
+                match ident_type {
+                    IdentifierType::TraceColumns(trace_columns) => {
+                        validate_access(i, trace_columns.size())?;
+                        Ok(Expression::NamedTraceAccess(NamedTraceAccess::new(
+                            ident.clone(),
+                            range.start() + i,
+                            named_trace_access.row_offset(),
+                        )))
+                    }
+                    _ => Err(SemanticError::InvalidListComprehension(format!(
+                        "Iterable {ident} should be trace columns"
+                    )))?,
```
nit: this grammar doesn't sound quite right. Should this be "Iterable {ident} should be a trace column" or "Iterable {ident} should contain trace columns"?

grjte reviewed (approved):
LGTM, great work @tohrnii, thanks! I left 2 final non-blocking nits inline

bobbinth reviewed (approved):
All looks good! Thank you!
